# Feature Selection Analysis

In [74]:
import pandas as pd
import numpy as np

In [75]:
fe_df=pd.read_csv('../../data/raw/incident_event_log.csv')

In [76]:
check_columns=['assignment_group','assigned_to','subcategory','impact','urgency','vendor','problem_id']

(fe_df[check_columns]=='?').sum()


assignment_group     14213
assigned_to          27496
subcategory            111
impact                   0
urgency                  0
vendor              141468
problem_id          139417
dtype: int64

In [77]:
(((fe_df[check_columns]=='?').sum()/len(fe_df))*100).round(2)


assignment_group    10.03
assigned_to         19.40
subcategory          0.08
impact               0.00
urgency              0.00
vendor              99.83
problem_id          98.38
dtype: float64

In [78]:

creation_records = fe_df[fe_df['sys_mod_count'] == 0]


target_cols = ['assignment_group', 'assigned_to', 'subcategory', 'impact', 'urgency', 'vendor', 'problem_id']


(creation_records[target_cols] == '?').sum()


assignment_group      906
assigned_to          7137
subcategory            69
impact                  0
urgency                 0
vendor              24898
problem_id          24673
dtype: int64

In [79]:
100-(((creation_records[target_cols] == '?').sum()/len(creation_records))*100).round(2)

assignment_group     96.36
assigned_to          71.35
subcategory          99.72
impact              100.00
urgency             100.00
vendor                0.06
problem_id            0.96
dtype: float64

In [80]:
fe_df['number'].value_counts()

number
INC0019396    58
INC0044260    56
INC0005927    46
INC0020718    45
INC0011206    44
              ..
INC0009063     2
INC0009060     2
INC0009055     2
INC0000686     2
INC0000682     2
Name: count, Length: 24918, dtype: int64

In [81]:
print('caller_id :',fe_df['caller_id'].value_counts().shape[0])
print("opened_by :",fe_df['opened_by'].value_counts().shape[0])
print('location :',fe_df['location'].value_counts().shape[0])
print('category :',fe_df['category'].value_counts().shape[0])
print('subcategory :',fe_df['subcategory'].value_counts().shape[0])
print('u_symptom :',fe_df['u_symptom'].value_counts().shape[0])
print('assignment_group :',fe_df['assignment_group'].value_counts().shape[0])
print('assigned_to :',fe_df['assigned_to'].value_counts().shape[0])


caller_id : 5245
opened_by : 208
location : 225
category : 59
subcategory : 255
u_symptom : 526
assignment_group : 79
assigned_to : 235


In [82]:
fe_df['caller_id'].value_counts().sort_values(ascending=False).head(10)

caller_id
Caller 1904    1425
Caller 290      791
Caller 4514     716
Caller 1441     322
Caller 298      293
Caller 3763     270
Caller 93       239
Caller 1531     231
Caller 4414     224
Caller 3160     220
Name: count, dtype: int64

In [83]:
fe_df['opened_by'].value_counts().sort_values(ascending=False).head(10)

opened_by
Opened by  17     41466
Opened by  24      8063
Opened by  131     7269
Opened by  108     6121
Opened by  55      5655
Opened by  40      4915
?                  4835
Opened by  62      4821
Opened by  59      4657
Opened by  501     4178
Name: count, dtype: int64

In [84]:
fe_df['assigned_to'].value_counts().sort_values(ascending=False).head(10)

assigned_to
?               27496
Resolver 17     10590
Resolver 13      8802
Resolver 194     3515
Resolver 57      3144
Resolver 69      2983
Resolver 33      2502
Resolver 115     2229
Resolver 24      2174
Resolver 249     2128
Name: count, dtype: int64

In [85]:
fe_df['caller_id'].value_counts().median()

np.float64(16.0)

In [86]:
fe_df['opened_by'].value_counts().median()

np.float64(33.5)

In [87]:
fe_df['assigned_to'].value_counts().median()

np.float64(163.0)

In [88]:
(fe_df['caller_id'].value_counts()==1).sum()

np.int64(0)

In [89]:
(fe_df['opened_by'].value_counts()==1).sum()

np.int64(0)

In [90]:
(fe_df['assigned_to'].value_counts()==1).sum()

np.int64(6)

In [91]:
(fe_df['caller_id'].value_counts()<5).sum()

np.int64(847)

In [92]:
(fe_df['opened_by'].value_counts()<5).sum()

np.int64(18)

In [93]:
(fe_df['assigned_to'].value_counts()<5).sum()

np.int64(19)

In [94]:
fe_df['impact'].value_counts()

impact
2 - Medium    134335
3 - Low         3886
1 - High        3491
Name: count, dtype: int64

In [95]:
fe_df['urgency'].value_counts()

urgency
2 - Medium    134094
1 - High        4020
3 - Low         3598
Name: count, dtype: int64

In [96]:
fe_df['priority'].value_counts()

priority
3 - Moderate    132452
4 - Low           4030
2 - High          2972
1 - Critical      2258
Name: count, dtype: int64

In [97]:
fe_df[['impact','urgency','priority']].value_counts()

impact      urgency     priority    
2 - Medium  2 - Medium  3 - Moderate    132429
3 - Low     3 - Low     4 - Low           3431
1 - High    1 - High    1 - Critical      2258
2 - Medium  1 - High    2 - High          1739
1 - High    2 - Medium  2 - High          1233
3 - Low     2 - Medium  4 - Low            432
2 - Medium  3 - Low     4 - Low            167
3 - Low     1 - High    3 - Moderate        23
Name: count, dtype: int64

In [98]:
print("incident_state has feature:",fe_df['incident_state'].unique().shape[0],"values")
print("sys_created_by has feature:",fe_df['sys_created_by'].unique().shape[0],"values")
print("sys_created_at has feature:",fe_df['sys_created_at'].unique().shape[0],"values")
print("notify has feature:",fe_df['notify'].unique().shape[0],"values")
print("")
print("Percentage of unknown values in incident_state  feature:",((((fe_df['incident_state'])=='?').sum()/len(fe_df))*100).round(2))
print("Percentage of unknown values in sys_created_by  feature:",((((fe_df['sys_created_by'])=='?').sum()/len(fe_df))*100).round(2))
print("Percentage of unknown values in sys_created_at  feature:",((((fe_df['sys_created_at'])=='?').sum()/len(fe_df))*100).round(2))
print("Percentage of unknown values in notify  feature:",((((fe_df['notify'])=='?').sum()/len(fe_df))*100).round(2))
print()

print(fe_df[['incident_state','sys_created_by','sys_created_at','notify']].describe())

incident_state has feature: 9 values
sys_created_by has feature: 186 values
sys_created_at has feature: 11553 values
notify has feature: 2 values

Percentage of unknown values in incident_state  feature: 0.0
Percentage of unknown values in sys_created_by  feature: 37.45
Percentage of unknown values in sys_created_at  feature: 37.45
Percentage of unknown values in notify  feature: 0.0

       incident_state sys_created_by sys_created_at         notify
count          141712         141712         141712         141712
unique              9            186          11553              2
top            Active              ?              ?  Do Not Notify
freq            38716          53076          53076         141593


In [99]:
((fe_df['contact_type'].value_counts()/len(fe_df))*100).round(2)

contact_type
Phone             99.12
Self service       0.70
Email              0.16
IVR                0.01
Direct opening     0.01
Name: count, dtype: float64

 # Feature Engineering Roadmap

## Input Features
contact_type
location
category
subcategory
u_symptom

assignment_group
assigned_to

caller_id
opened_by

impact
urgency
priority

Hour
Day_of_Week
Month

## Target
made_sla

In [100]:
%store -r eda_df

In [101]:
feature_df=eda_df[['contact_type','location','category','subcategory','u_symptom','assignment_group','assigned_to','caller_id','opened_by','impact','urgency','priority','Hour','Day_of_week','Month','made_sla']]
feature_df.columns.shape

(16,)

In [102]:
X=feature_df.drop(['made_sla'],axis=1)
X.shape

(141712, 15)

In [103]:
y=feature_df['made_sla']
y.shape

(141712,)

In [104]:
from sklearn.model_selection import train_test_split

In [105]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42,stratify=y)

In [106]:
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((113369, 15), (28343, 15), (113369,), (28343,))

In [115]:
X_train.to_csv('../../data/processed/X_train.csv',index=False)
X_test.to_csv('../../data/processed/X_test.csv',index=False)
y_train.to_csv('../../data/processed/y_train.csv',index=False)
y_test.to_csv('../../data/processed/y_test.csv',index=False)

### Train-Test Split Validation

The dataset was split using Stratified Sampling to preserve the original target distribution.

Train Records : 113,369
Test Records  : 28,343

Target distribution will be validated to ensure no class imbalance drift occurred during splitting.

In [107]:
((y_train.value_counts()/len(y_train))*100).round(2)

made_sla
True     93.5
False     6.5
Name: count, dtype: float64

In [108]:
((y_test.value_counts()/len(y_test))*100).round(2)

made_sla
True     93.5
False     6.5
Name: count, dtype: float64

In [109]:
print(y.value_counts(normalize=True,))

print(y_train.value_counts(normalize=True))

print(y_test.value_counts(normalize=True))

made_sla
True     0.934974
False    0.065026
Name: proportion, dtype: float64
made_sla
True     0.934973
False    0.065027
Name: proportion, dtype: float64
made_sla
True     0.934975
False    0.065025
Name: proportion, dtype: float64


### Train-Test Split Validation

The dataset was split using stratified sampling to preserve the target class distribution.

Original Dataset:
- SLA Met: 93.50%
- SLA Breach: 6.50%

Training Dataset:
- SLA Met: 93.50%
- SLA Breach: 6.50%

Testing Dataset:
- SLA Met: 93.50%
- SLA Breach: 6.50%

The class distribution remains consistent across train and test datasets, confirming that stratified sampling successfully preserved the target imbalance characteristics and ensuring reliable model evaluation.

## Missing Value Analysis (Training Dataset)

Missing value analysis was performed exclusively on the training dataset to avoid introducing information from the test dataset into feature engineering decisions.

The symbol '?' is treated as an unknown category rather than a traditional missing value because it may carry business meaning in the incident management process.

In [110]:
X_train_missing=(X_train[['contact_type','location','category','subcategory','u_symptom','assignment_group','assigned_to','caller_id','opened_by']]=='?').sum()
X_train_missing=X_train_missing.reset_index()
X_train_missing.columns=['features','Missing Count (?)']
X_train_missing['Missing %']=((X_train_missing['Missing Count (?)']/len(X_train))*100).round(2)
X_train_missing

,features,Missing Count (?),Missing %
0,contact_type,0,0.00
1,location,56,0.05
2,category,63,0.06
3,subcategory,86,0.08
4,u_symptom,26378,23.27
5,assignment_group,11420,10.07
6,assigned_to,22028,19.43
7,caller_id,24,0.02
8,opened_by,3867,3.41


## Missing Value Treatment Strategy

| Feature          | Replacement           |
| ---------------- | --------------------- |
| location         | Unknown_Location      |
| category         | Unknown_Category      |
| subcategory      | Unknown_Subcategory   |
| u_symptom        | Unknown_Symptom       |
| assignment_group | Not_Assigned_Group    |
| assigned_to      | Not_Assigned_Engineer |
| caller_id        | Unknown_Caller        |
| opened_by        | Unknown_Reporter      |


In [111]:
X_train['impact'].unique(),X_train['urgency'].unique(),X_train['priority'].unique(),

(array(['2 - Medium', '1 - High', '3 - Low'], dtype=object),
 array(['2 - Medium', '1 - High', '3 - Low'], dtype=object),
 array(['3 - Moderate', '1 - Critical', '2 - High', '4 - Low'],
       dtype=object))

## Ordinal Feature Engineering

Impact, urgency and priority contain an inherent business hierarchy.

Unlike nominal categorical variables, the order between categories carries meaningful information.

Therefore, ordinal encoding was selected to preserve the severity relationship while converting categories into numerical representations suitable for machine learning models.

### Ordinal Mapping Strategy

Impact:
- Low → 1
- Medium → 2
- High → 3

Urgency:
- Low → 1
- Medium → 2
- High → 3

Priority:
- Low → 1
- Moderate → 2
- High → 3
- Critical → 4

This mapping ensures that higher numerical values correspond to higher business severity, improving interpretability during model evaluation and SHAP analysis.

## Feature Type and Encoding Summary

| Feature          | Feature Type       | Treatment          |
| ---------------- | ------------------ | ------------------ |
| contact_type     | Categorical        | Target Encoding    |
| location         | Categorical        | Target Encoding    |
| category         | Categorical        | Frequency Encoding |
| subcategory      | Categorical        | Frequency Encoding |
| u_symptom        | High Cardinality   | Target Encoding    |
| assignment_group | Medium Cardinality | Frequency Encoding |
| assigned_to      | High Cardinality   | Frequency Encoding |
| caller_id        | High Cardinality   | Target Encoding    |
| opened_by        | Medium Cardinality | Frequency Encoding |
| impact           | Ordinal            | Ordinal Encoding   |
| urgency          | Ordinal            | Ordinal Encoding   |
| priority         | Ordinal            | Ordinal Encoding   |
| Hour             | Numerical          | Pass Through       |
| Day_of_week      | Calendar Categorical | CatBoost Categorical Feature |
| Month       | Calendar Categorical | CatBoost Categorical Feature               |


## Data Leakage Prevention

To ensure realistic model evaluation:

1. Feature selection was completed before modeling.
2. Train-test split was performed before any encoding.
3. Frequency encoding mappings will be learned from training data only.
4. Target encoding mappings will be learned from training data only.
5. Unseen categories in test/production data will use the global target mean.
6. SMOTE (if used) will be applied only on training data.
7. Calendar features (Month and Day_of_week) will be treated as categorical variables and transformed within the training pipeline.

## Feature Engineering Conclusion
Feature engineering decisions were finalized based on business understanding, EDA findings, cardinality analysis, missing value analysis and leakage prevention principles. The resulting feature set will be transformed through a preprocessing pipeline in the modeling phase. The implementation of preprocessing, encoding, imbalance handling and model training will be performed in 05_model_training.ipynb using Scikit-Learn Pipelines and ColumnTransformer.
All learned transformations such as frequency encoding and target encoding will be fitted exclusively on the training dataset to avoid data leakage.